File: python/pyrgbd/file_writer_helper.py
from .camera_calibration import CameraCalibration
from .depth_decoder import DepthCodecType
from .file import FileVideoFrame, FileAudioFrame, FileIMUFrame, FileTRSFrame
from .file_writer import NativeFileWriterConfig, NativeFileWriter
from .yuv_frame import YuvFrame
from ._librgbd_ffi import lib


class NativeFileWriterHelper:
    def __init__(self):
        self.ptr = lib.rgbd_file_writer_helper_ctor()

    def close(self):
        lib.rgbd_file_writer_helper_dtor(self.ptr)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def set_calibration(self, calibration: CameraCalibration):
        with calibration.to_native() as native_calibration:
            lib.rgbd_file_writer_helper_set_calibration(self.ptr, native_calibration.ptr)

    def set_depth_codec_type(self, depth_codec_type: DepthCodecType):
        lib.rgbd_file_writer_helper_set_depth_codec_type(self.ptr, depth_codec_type)

    def set_depth_unit(self, depth_unit: float):
        lib.rgbd_file_writer_helper_set_depth_unit(self.ptr, depth_unit)

    def set_cover(self, cover: YuvFrame):
        with cover.to_native() as native_cover:
            lib.rgbd_file_writer_helper_set_cover(self.ptr, native_cover.ptr)

    def add_video_frame(self, video_frame: FileVideoFrame):
        with video_frame.to_native() as native_video_frame:
            lib.rgbd_file_writer_helper_add_video_frame(self.ptr, native_video_frame.ptr)

    def add_audio_frame(self, audio_frame: FileAudioFrame):
        with audio_frame.to_native() as native_audio_frame:
            lib.rgbd_file_writer_helper_add_audio_frame(self.ptr, native_audio_frame.ptr)

    def add_imu_frame(self, imu_frame: FileIMUFrame):
        with imu_frame.to_native() as native_imu_frame:
            lib.rgbd_file_writer_helper_add_imu_frame(self.ptr, native_imu_frame.ptr)

    def add_trs_frame(self, trs_frame: FileTRSFrame):
        with trs_frame.to_native() as native_trs_frame:
            lib.rgbd_file_writer_helper_add_trs_frame(self.ptr, native_trs_frame.ptr)

    def write_to_path(self, path: str):
        lib.rgbd_file_writer_helper_write_to_path(self.ptr, path.encode("utf8"))

File: web/rgbd-js/file_writer_helper.ts
import { CameraCalibration } from "./camera_calibration";
import { NativeByteArray } from "./capi_containers";
import { DepthCodecType } from "./depth_decoder";
import { FileVideoFrame, FileAudioFrame, FileIMUFrame, FileTRSFrame } from "./file";
import { YuvFrame } from "./yuv_frame";

export class NativeFileWriterHelper {
  wasmModule: any;
  ptr: number;

  constructor(wasmModule: any) {
    this.wasmModule = wasmModule;
    this.ptr = this.wasmModule.ccall('rgbd_file_writer_helper_ctor', 'number', [], []);
  }

  close() {
    this.wasmModule.ccall('rgbd_file_writer_helper_dtor', null, ['number'], [this.ptr]);
  }

  setCalibration(calibration: CameraCalibration) {
    const nativeCalibration = calibration.toNative(this.wasmModule);
    this.wasmModule.ccall('rgbd_file_writer_helper_set_calibration', null, ['number', 'number'], [this.ptr, nativeCalibration.ptr]);
    nativeCalibration.close();
  }

  setDepthCodecType(depthCodecType: DepthCodecType) {
    this.wasmModule.ccall('rgbd_file_writer_helper_set_depth_codec_type', null, ['number', 'number'], [this.ptr, depthCodecType]);
  }

  setDepthUnit(depthUnit: number) {
    this.wasmModule.ccall('rgbd_file_writer_helper_set_depth_unit', null, ['number', 'number'], [this.ptr, depthUnit]);
  }

  setCover(cover: YuvFrame) {
    const nativeCover = cover.toNative(this.wasmModule);
    this.wasmModule.ccall('rgbd_file_writer_helper_set_cover', null, ['number', 'number'], [this.ptr, nativeCover.ptr]);
    nativeCover.close();
  }

  addVideoFrame(videoFrame: FileVideoFrame) {
    const nativeVideoFrame = videoFrame.toNative(this.wasmModule);
    this.wasmModule.ccall('rgbd_file_writer_helper_add_video_frame', null, ['number', 'number'], [this.ptr, nativeVideoFrame.ptr]);
    nativeVideoFrame.close();
  }

  addAudioFrame(audioFrame: FileAudioFrame) {
    const nativeAudioFrame = audioFrame.toNative(this.wasmModule);
    this.wasmModule.ccall('rgbd_file_writer_helper_add_audio_frame', null, ['number', 'number'], [this.ptr, nativeAudioFrame.ptr]);
    nativeAudioFrame.close();
  }

  addIMUFrame(imuFrame: FileIMUFrame) {
    const nativeIMUFrame = imuFrame.toNative(this.wasmModule);
    this.wasmModule.ccall('rgbd_file_writer_helper_add_imu_frame', null, ['number', 'number'], [this.ptr, nativeIMUFrame.ptr]);
    nativeIMUFrame.close();
  }

  addTRSFrame(trsFrame: FileTRSFrame) {
    const nativeTRSFrame = trsFrame.toNative(this.wasmModule);
    this.wasmModule.ccall('rgbd_file_writer_helper_add_trs_frame', null, ['number', 'number'], [this.ptr, nativeTRSFrame.ptr]);
    nativeTRSFrame.close();
  }

  writeToBytes(): Uint8Array {
    const nativeBytesPtr = this.wasmModule.ccall('rgbd_file_writer_helper_write_to_bytes', null, ['number'], [this.ptr]);
    const nativeBytes = new NativeByteArray(this.wasmModule, nativeBytesPtr);
    const bytes = nativeBytes.toArray();
    nativeBytes.close();
    return bytes;
  }
}